### PR TITLE
[QA-1932] Terra-UI testing: Handle NPS survey popup if found

### PR DESCRIPTION
**Pull request by aweng98**
[QA-1932](https://broadworkbench.atlassian.net/browse/QA-1932)

Unexpected NPS survey popup cause UI test to fail.

Screen Shot 2022-06-27 at 7 20 26 PM

### CODE REVIEW

File: integration-tests/utils/integration-helpers.js
Comment (aweng98) [inline integration-tests/utils/integration-helpers.js:205]: Call `signIntoTerra()` first to handle unexpected NPS survey popup.

File: integration-tests/utils/integration-utils.js
Comment (aweng98) [inline integration-tests/utils/integration-utils.js]: Tried to use text like this but unable to make it work. 
`await iframe.$x(clickable({text: "Ask Me Later"}))`.

File: integration-tests/utils/integration-helpers.js
Comment (cahrens) [inline integration-tests/utils/integration-helpers.js:205]: Was there a reason to switch this order?

File: integration-tests/utils/integration-utils.js
Comment (cahrens) [inline integration-tests/utils/integration-utils.js]: Can you use the user-visible text instead of an ID?

File: integration-tests/utils/integration-helpers.js
Comment (aweng98) [reply integration-tests/utils/integration-helpers.js:205]: See my comment for this change.

Comment (aweng98) [reply integration-tests/utils/integration-helpers.js:205]: You're too fast! 🙂 I was still typing comments while you were reviewing it.

Comment (cahrens) [reply integration-tests/utils/integration-helpers.js:205]: Ah right, you are doing the dismissal within signIntoTerra. Can you add a comment in the code? I think the original idea was that "View Workspaces" was sort of an entryway into logging into Terra, so people might be tempted to swap the order without a comment.

File: integration-tests/utils/integration-utils.js
Comment (aweng98) [inline integration-tests/utils/integration-utils.js]: Changed how click works here because `elementHandle.click()` uses element's (x, y) coordinates. If there are more than one notifications, after one notification is closed, position of next notification moves. So click by coordinates is bad idea here as closing additional notifications can unexpectedly fail.

**Comment by nawatts** [comment]
Why do we even show the NPS survey on non-prod environments? Can it be turned off instead of adding code to handle it?

Also, it looks like there are a lot of changes in this PR unrelated to the NPS survey. What motivated those changes?

File: integration-tests/utils/integration-helpers.js
Comment (aweng98) [reply integration-tests/utils/integration-helpers.js:205]: Good idea. Added comment: `// Sign in to handle unexpected NPS survey popup and Loading Terra... spinner`

**Comment by aweng98** [comment]
> Also, it looks like there are a lot of changes in this PR unrelated to the NPS survey.

"Unrelated" changes are `dismissNotifications()` and `waitForNoSpinners()`, right? I discovered them as flaky while I was testing new code for NPS survey. If you feel strongly about they don't belong here, I can remove them. I don't have a ticket for them alone. I thought that was going to be brought up as it is kind of "controversial".

**Comment by aweng98** [comment]
> Why do we even show the NPS survey on non-prod environments? Can it be turned off instead of adding code to handle it?

That's a good question for PM and team that owns the functionality.

File: integration-tests/utils/integration-utils.js
Comment (nawatts) [reply integration-tests/utils/integration-utils.js]: It seems like this could be simplified to
```js
notificationCloseButtons.map(button => button.evaluate(btn => btn.click()))
```
`evaluate` and `evaluateHandle` automatically pass the handle as the first argument to the function, so passing `button` again is unnecessary.

https://pptr.dev/#?product=Puppeteer&version=v15.1.1&show=api-elementhandleevaluatepagefunction-args

File: integration-tests/utils/integration-utils.js
Comment (nawatts) [inline integration-tests/utils/integration-utils.js:209]: Why does this error need to be caught and logged here? Would it not be be logged by the `page.pageerror` handler attached in `enablePageLogging`?

File: integration-tests/utils/integration-utils.js
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: It looks like this could be simplified to
```js
await closeButton.evaluate(button => button.click())
```

`evaluate` and `evaluateHandle` automatically pass the handle as the first argument to the function. So there's no need to pass `closeButton` again.
https://pptr.dev/#?product=Puppeteer&version=v15.1.1&show=api-elementhandleevaluatepagefunction-args

File: integration-tests/utils/integration-utils.js
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: Why do we need to wait for the root element? The root element is part of the initial HTML response, so if the page is loaded, it should be present.

**Comment by aweng98** [comment]
To avoid back-and-forth comments, I've reverted `dismissNotifications()` and `waitForNoSpinners()` changes.

File: integration-tests/utils/integration-utils.js
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: Why do we need to wait the root element to be present? It's part of the HTML response, so if the page is loaded, it should be present.

File: integration-tests/utils/integration-utils.js
Comment (nawatts) [inline integration-tests/utils/integration-utils.js]: Why change the selector here?

**Comment by nawatts** [comment]
> "Unrelated" changes are dismissNotifications() and waitForNoSpinners(), right? I discovered them as flaky while I was testing new code for NPS survey. If you feel strongly about they don't belong here, I can remove them. I don't have a ticket for them alone. I thought that was going to be brought up as it is kind of "controversial".

I think it's fine include a few small semi-related changes, but they should be explained in the PR description or comments. Also, grouping multiple changes in one PR blocks merging any of those changes until all of them are approved. So if some change is likely to generate discussion, it's likely worth separating that into a smaller, focused PR.

File: integration-tests/utils/integration-utils.js
Comment (cahrens) [reply integration-tests/utils/integration-utils.js]: That might be because the button is an iframe. Did you try `iframe.$x('.//*[normalize-space(.)="Ask Me Later"]')? In other words, build up the xPath using the text.

File: integration-tests/utils/integration-utils.js
Comment (aweng98) [reply integration-tests/utils/integration-utils.js]: Nice! Done. I didn't know this before.

**Comment by cahrens** [comment]
> > Why do we even show the NPS survey on non-prod environments? Can it be turned off instead of adding code to handle it?
> 
> That's a good question for PM and team that owns the functionality.

I posted in the UX channel to ask about this-- https://broadinstitute.slack.com/archives/C34GAKQ5R/p1656435309063549

File: integration-tests/utils/integration-utils.js
Comment (aweng98) [reply integration-tests/utils/integration-utils.js:209]: Log here because it would be easy to find them in (middle of somewhere) test log file instead finding them at the end of log file in test Summary.
For the second question, I don't know if error would be caught by the `page.pageerror`.

**Review by aweng98 — COMMENTED** [review]
Hi, Nick. I have reverted unrelated improvement in this PR.

File: integration-tests/utils/integration-utils.js
Comment (nawatts) [reply integration-tests/utils/integration-utils.js:209]: Doesn't the `page.pageerror` handler log errors when they occur? Why would this message be in a different place in the logs than the "dismissing NPS survey" message?

Comment (aweng98) [reply integration-tests/utils/integration-utils.js:209]: I think you're specifically talking about any error from clicking, right? I still don't know if `page.pageerror` will catch click error.

This catch block also catches error if the button in iframe cannot not be found. Other type of errors are not going to be captured by `page.pageerror`.

Comment (nawatts) [reply integration-tests/utils/integration-utils.js:209]: Oh yeah, I'm getting mixed up. `page.pageerror` wouldn't catch this. Though I'm still not convinced there's a need to log this separately vs letting the usual error handling machinery log it.

**Review by nawatts — APPROVED** [review]

File: integration-tests/utils/integration-utils.js
Comment (aweng98) [reply integration-tests/utils/integration-utils.js]: Sorry for the delayed response. Yes, that xpath worked! It took me some time to run tests repeatly to make sure proposed changes works.

**Review by cahrens — APPROVED** [review]
It seems that Joy might be turning off the NPS Survey, but I don't see harm in checking in the code to handle it when it comes back.

**Comment by aweng98** [comment]
Thank you all.

**Pull request merged**
